### PR TITLE
fix newapp resource convert for extensions and apps groups

### PR DESCRIPTION
**Pull request by juanvallejo**
Fixes https://github.com/openshift/origin/issues/15783

**Before**
```
$ oc new-app -f examples/prometheus/prometheus.yaml -o yaml
error: extensions.Deployment is not suitable for converting to ""
```

**After**
```
$ oc new-app -f examples/prometheus/prometheus.yaml -o yaml | head
apiVersion: v1
items:
- apiVersion: v1
  kind: ServiceAccount
  metadata:
    annotations:
      openshift.io/generated-by: OpenShiftNewApp
      serviceaccounts.openshift.io/oauth-redirectreference.primary: '{"kind":"OAuthRedirectReference","apiVersion":"v1","reference":{"kind":"Route","name":"prometheus"}}'
    creationTimestamp: null
    name: prometheus
```

cc @fabianofranz @smarterclayton

### CODE REVIEW

File: pkg/cmd/util/cmd.go [hunk 1]
```diff
@@ -121,6 +133,18 @@ func convertItemsForDisplay(objs []runtime.Object, preferredVersions ...schema.G
 			}
 		}
 
+		// if no preferred version found in the list of given GroupVersions,
+		// attempt to convert to first GroupVersion that satisfies a preferred version
+		if len(actualOutputVersion.Version) == 0 {
+			for _, version := range preferredVersions {
```
Comment (smarterclayton) [inline pkg/cmd/util/cmd.go]: You shouldn't even need to do this loop - you should be passing preferredVersions in to ConvertToVersion as a Versioner.

File: pkg/cmd/util/cmd.go [hunk 2]
```diff
@@ -95,6 +95,18 @@ func convertItemsForDisplay(objs []runtime.Object, preferredVersions ...schema.G
 			groups = append(groups, groupMeta)
 		}
 
+		// if no preferred versions given, convert item to first group version found.
+		// the given object is guaranteed to be at least part of one group by this point.
+		if len(preferredVersions) == 0 {
+			defaultGroupVersion := groups[0].GroupVersion
+			convertedObject, err := kapi.Scheme.ConvertToVersion(obj, defaultGroupVersion)
```
Comment (smarterclayton) [inline pkg/cmd/util/cmd.go]: You should just be passing in groupVersions to ConvertToVersion (you'll need to cast to the runtime schema type which handles Versioner).

**Comment by smarterclayton** [comment]
Tests.

**Comment by juanvallejo** [comment]
@smarterclayton thanks for the feedback, review comments addressed

File: test/cmd/newapp.sh [hunk 3]
```diff
@@ -95,6 +95,14 @@ os::cmd::expect_success_and_text 'oc new-app ruby-helloworld-sample --param MYSQ
 os::cmd::expect_success_and_text  'oc new-app -e FOO=BAR -f examples/jenkins/jenkins-ephemeral-template.json -o jsonpath="{.items[?(@.kind==\"DeploymentConfig\")].spec.template.spec.containers[0].env[?(@.name==\"FOO\")].value}" ' '^BAR$'
 os::cmd::expect_success_and_text  'oc new-app -e OPENSHIFT_ENABLE_OAUTH=false -f examples/jenkins/jenkins-ephemeral-template.json -o jsonpath="{.items[?(@.kind==\"DeploymentConfig\")].spec.template.spec.containers[0].env[?(@.name==\"OPENSHIFT_ENABLE_OAUTH\")].value}" ' 'false'
 
+# check that multiple resource groups are printed with their respective external version
+os::cmd::expect_success_and_text 'oc new-app -f examples/prometheus/prometheus.yaml -o yaml' 'apiVersion: apps/v1beta1'
+# check that if an --output-version is requested on a list of varying resource kinds, an error is returned if
+# at least one of the resource groups does not support the given version
+os::cmd::expect_failure_and_text 'oc new-app -f examples/prometheus/prometheus.yaml -o yaml --output-version=v1' 'extensions.Deployment is not suitable for converting'
```
Comment (fabianofranz) [inline test/cmd/newapp.sh]: Minor comment, make it more defensive against future changes by just checking "not suitable for converting".

Comment (juanvallejo) [inline test/cmd/newapp.sh]: Done!

**Comment by juanvallejo** [comment]
@fabianofranz 
> Minor comment, make it more defensive against future changes by just checking "not suitable for converting".

Since `examples/prometheus/prometheus.yaml` contains multiple resources of different group versions, any time I test the `--output-version` flag I will get at least one resource that is incompatible with the user-supplied version (e.g. `--output-version=extensions/v1beta1` would not be compatible with `api.Secret`). I have gone ahead and created a new test file that only deals with two specific resources (`api.Secret`, and `extensions.Deployment`) in order to check for the specific resource in the tests.

**Pull request merged**